In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

**--> Reading and analysing the dataset**

In [11]:
credit_card_data=pd.read_csv('/content/CreditCardDefault.csv')

In [12]:
# first 5 lines of dataset
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [13]:
# last 5 lines of dataset
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [14]:
# information about dataset
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [16]:
# check missing values 
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [29]:
# check distribution of legit and fraud transaction
credit_card_data.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

**Dataset is highly unbalanced as legit transactions >>> fraud/default transaction** 

0 --> legit transaction

1 --> fraud transaction

In [31]:
# separating data into legit and fraud dataset

legit=credit_card_data[credit_card_data.Class ==0]
fraud=credit_card_data[credit_card_data.Class ==1]

In [22]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [27]:
# analysing legit and fraud data

legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [32]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [35]:
# compare the mean values for both the transactions

credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

**Build another sample dataset containing similar distribution of legit and fraud transactions**

fraud transactions --> 492 , 
 legit transactions --> 284315


**Take 492 transactions randomly from legit transactions and then combine it with fraud transactions , so that distribution become balanced**

In [36]:
legit_sample=legit.sample(n=492)

In [88]:
# concatenating fraud and legit_sample datasets
new_dataset=pd.concat([legit_sample,fraud]) # our ml model will be trained on this dataset
new_dataset.shape

(984, 31)

In [74]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
167858  118932.0 -0.582434  0.613298 -0.616204 -0.912547  0.984297 -0.960111   
87126    61546.0 -1.103298 -0.701083  0.158903 -2.796331  2.037790  3.321275   
224483  143815.0 -0.730206  1.329831  0.779536  0.750508  0.991511 -0.857806   
278383  168193.0 -0.712467  0.928535 -0.589258 -1.852311  0.878231 -0.312251   
229510  145953.0  2.072710  0.195039 -1.758256  0.378240  0.562975 -0.768235   

              V7        V8        V9  ...       V21       V22       V23  \
167858  0.896459 -0.120062  0.222765  ... -0.147671 -0.174413  0.151759   
87126  -0.275242  0.952714  1.322803  ...  0.170823  0.717980  0.365689   
224483  1.893707 -0.683607 -1.036929  ... -0.006474  0.195653 -0.451227   
278383  1.184719 -0.426181  1.187671  ... -0.115209 -0.017994  0.000506   
229510  0.251574 -0.278475  0.370528  ... -0.360396 -0.905140  0.303545   

             V24       V25       V26       V27       V28  Amount  Class  
167858 -0.780528 -0.095276 -0.135605 -0.448920 -0.323001    0.00      0  
87126   1.019198  0.172424 -0.729660  0.195237  0.068878   65.45      0  
224483  0.044354  0.958365 -0.337682 -0.337538 -0.212472   39.00      0  
278383  0.014828 -0.574935 -1.366761 -0.841157 -0.233973   31.08      0  
229510  0.403397 -0.210109  0.177626 -0.058511 -0.028887    8.99      0  

[5 rows x 31 columns]

In [75]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [76]:
new_dataset.Class.value_counts()

0    492
1    492
Name: Class, dtype: int64

In [77]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92002.896341  0.002473  0.050901  0.035052 -0.042205 -0.027034   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.041651 -0.016254 -0.040899  0.002228  ...  0.011885 -0.049712   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.040618  0.021210 -0.016483 -0.012393  0.016606 -0.022094  0.006619   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       82.942500  
1      122.211321  

[2 rows x 30 columns]

In [78]:
new_dataset.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [79]:
# separating features and targets
X=new_dataset[['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount']]
y=new_dataset['Class']

In [80]:
print(X.shape)
print(y.shape)

(984, 30)
(984,)


In [81]:
# train-test split
X_train,X_test,y_train,y_test=train_test_split(X,y,train_size=0.8,stratify=y,random_state=2)

In [82]:
print(X.shape,X_train.shape,X_test.shape)

(984, 30) (787, 30) (197, 30)


**Model training**

In [83]:
# train machine learning model with training data
model=LogisticRegression()
model.fit(X_train,y_train)

LogisticRegression()

**Evaluation of model**

In [84]:
# calculating accuracy on training data

X_train_prediction=model.predict(X_train)
training_data_accuracy= accuracy_score(X_train_prediction,y_train)

In [89]:
print('Accuracy on training data : ',training_data_accuracy)

Accuracy on training data :  0.9174078780177891


In [86]:
# calculating accuracy on test data

X_test_prediction=model.predict(X_test)
test_data_accuracy= accuracy_score(X_test_prediction,y_test)

In [87]:
print('Accuracy on test data : ',test_data_accuracy)


Accuracy on test data :  0.883248730964467
